src/effects/native/gatereffect.cpp

```diff
@@ -0,0 +1,236 @@
+{
+    Q_UNUSED(manifest);
+}
+
+GaterEffect::~GaterEffect() {
+}
+
+void GaterEffect::processChannel(const ChannelHandle& handle,
+                                GaterGroupState* pState,
+                                const CSAMPLE* pInput, CSAMPLE* pOutput,
+                                const unsigned int numSamples,
+                                const unsigned int sampleRate,
+                                const EffectProcessor::EnableState enableState,
+                                const GroupFeatureState& groupFeatures) { 
+                                    
+    Q_UNUSED(handle);
+                                    
+    const double rate  = m_pRateParameter->value();
+    const double shape  = m_pShapeParameter->value();
+
+    // Used to divide a beat by a power of 2
+    int divider = pow(2, floor(3*rate));
+
+    // Attack & Release increase with the divider factor
+    double attackInc = kBaseAttackInc*divider;
+    double releaseInc = kBaseReleaseInc*divider;
+
+    // Get channel specific state variable
+    unsigned int timePosition    = pState->timePosition;
+    unsigned int holdCounter     = pState->holdCounter;
+    double       gain            = pState->gain;
+    GaterGroupState::State state = pState->state;
+
+    // Use to keep track of the beginning of beats when playing with the rate
+    unsigned int beatPeriod = 1;
+    if (enableState == EffectProcessor::ENABLING) {        
+        // Start with the gate closed
+        state = GaterGroupState::IDLE;
+        gain = 0;
+
+        // Set initial time from current beat position, or 0 if no beat grid
+        if (groupFeatures.has_beat_length_sec && groupFeatures.has_beat_fraction) {
+            beatPeriod   = groupFeatures.beat_length_sec*sampleRate;
+            timePosition = groupFeatures.beat_fraction*beatPeriod;
+        } else {
+            timePosition = 0;   
+            beatPeriod   = sampleRate; // 1 second
+        }
+    }
+
+    // Adjust trigger period
+    unsigned int triggerPeriod;
+    double period = 1-rate;
+    if (groupFeatures.has_beat_length_sec) {
+        // triggerPeriod is a number of beats
+        if (m_pQuantizeParameter->toBool()) {
+            period = 1.0/divider;
+            
+            if (m_pTripletParameter->toBool()) {
+                period /= 3.0/4;
+            }
+        } else if (period < 1 / 8.0) {
+            period = 1 / 8.0;
+        }
+        triggerPeriod = period * groupFeatures.beat_length_sec * sampleRate;
+    } else {
+        // triggerPeriod is a number of seconds
+        period = std::max(period, 1 / 8.0);
+        triggerPeriod = period * sampleRate;
+    }
+
+    // triggerTime : offset for the gate trigger
+    unsigned int triggerTime;
+    if (m_pInvertParameter->toBool()) {
+        triggerTime = triggerPeriod/2;
+    } else {
+        triggerTime = 0;
+    }
+
+    // holdTime : number of samples spent in the Hold state
```

Comment:
Use variable names to clearly differentiate variables that are in units of samples versus time (seconds). For example, `holdTimeSamples` or `holdSamples`.